src/transformStateWithClones.js

```diff
@@ -7,7 +7,30 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const resultArray = [];
+  let clonedState = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(clonedState, action.extraData);
+        break;
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          if (clonedState[key]) {
+            delete clonedState[key];
+          }
+        }
+        break;
+      case 'clear':
+        clonedState = {};
+        break;
+    }
+
+    resultArray.push({ ...clonedState });
```

Comment:
The action of cloning the state is repeated in every case. Consider performing this action once after the `switch` statement to adhere to the DRY principle.